Add cache for ILS data

EreMaijala opened the pull request:
Caching ILS data makes it possible to avoid e.g. repeated patronLogin calls. 

Currently caching is enabled only for select few calls that return information that is not changing regularly. There are two separate caching mechanisms depending on whether the information is user-specific.

EreMaijala committed:
Add cache for ILS data

Caching ILS data makes it possible to avoid e.g. repeated patronLogin calls.

demiankatz commented on module/VuFind/src/VuFind/ILS/Connection.php:
```diff
@@ -1227,4 +1288,82 @@ public function __call($methodName, $params)
+    public function __call($methodName, $params)
+    {
+        if ($entry = $this->getCachedData($methodName, $params)) {
+            return $entry['data'];
```
In line 1313, you return `$entry['data']` and now here you're also referencing `$entry['data']`. I think this is correct and there really is an `$entry['data']['data']` in play... but that's a bit confusing. Might there be a way to make this a little easier to follow?

EreMaijala replied:
Yes, renamed keys and arrays, and added comments. Maybe better now?

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- this looks pretty reasonable based on a quick review, though I haven't tried anything hands-on yet. A couple of general thoughts:

1.) What target version did you have in mind for this? I'm thinking 11.0 might be more appropriate since this is a fairly substantial low-level change (even if configuration allows us to turn it off)... but if you strongly feel 10.1 is better, I may be persuadable. :-)

2.) Do you have any concerns about session size getting out of control if, for example, there's a huge amount of location data? Do we need any safety mechanisms on this?

EreMaijala committed:
Address issues.

EreMaijala commented:
@demiankatz 

1. I was thinking 10.1 since this will save a lot of calls and generally make things faster for many users.
2. True, I didn't really think about that. I removed pick up locations for now and added a comment. This can be revisited after [VUFIND-1652](https://openlibraryfoundation.atlassian.net/browse/VUFIND-1652).

EreMaijala committed:
Merge remote-tracking branch 'origin/dev' into dev-ils-cache

demiankatz committed:
Fix typo.

demiankatz commented on module/VuFind/src/VuFind/ILS/Connection.php:
```diff
@@ -1227,4 +1289,87 @@ public function __call($methodName, $params)
+        if (!$cacheLifeTime || !$cacheStorage) {
+            return null;
+        }
+        $cacheKey = $methodName . md5(serialize($params));
```
Would it be better to pass in the $cacheKey instead of the $params both here and in the put method? This way, you would only need to calculate the hash in one place, which would both improve efficiency (no double hashing/serialization needed) and reduce redundancy (no chance of the algorithm somehow getting out of sync due to code changes). It would also be possible to create a separate support method for generating the key, in case that would be a useful customization point.

demiankatz replied:
Or perhaps even better (additionally), the getCacheKey method could also contain the logic to determine whether caching is supported and return null if it is not. Then more logic could be centralized/deduplicated.

EreMaijala replied:
I refactored it a bit further. What do you think?

demiankatz reviewed (changes requested):
Thanks, @EreMaijala! This is looking pretty good to me. I went ahead and fixed a small typo in a comment. Beyond that, just one final thought....

EreMaijala committed:
Refactor.

demiankatz commented on module/VuFind/src/VuFind/ILS/Connection.php:
```diff
@@ -1227,4 +1289,97 @@ public function __call($methodName, $params)
+     *
+     * @return ?array
+     */
+    protected function getCacheSettings($methodName, $params): ?array
```
Worth adding types here?

```suggestion
    protected function getCacheSettings(string $methodName, array $params): ?array
```

EreMaijala replied:
I dared not enforce types here because the existing methods don't have typing either. I think e.g. $params might work if passed null even if you're not supposed to. Or do you think this would be a good moment to add types to the existing methods too?

EreMaijala replied:
Perhaps adding of types could be done in dev-11.0?

demiankatz replied:
Yes, I agree, if there's any risk of breaking something, the change should wait until 11.0.

demiankatz commented on module/VuFind/src/VuFind/ILS/Connection.php:
```diff
@@ -1227,4 +1289,97 @@ public function __call($methodName, $params)
+    {
+        $cacheKey = $cacheSettings['key'];
+        if ('shared' === $cacheSettings['storage']) {
+            $this->putSharedCachedData($cacheKey, $data);
```
Shouldn't you be accounting for the lifetime here? Otherwise I think the default 30 seconds will be used, unless I'm missing something!

EreMaijala replied:
Yikes! Thanks, for catching that. The call was missing the third parameter.

demiankatz reviewed (changes requested):
Thanks, @EreMaijala, I think this is much better, but I also think I've spotted a small bug.

EreMaijala committed:
Fix expiration handling.

EreMaijala committed:
Add support for default cache life time.

EreMaijala committed:
Tweak config.ini

EreMaijala committed:
Invalidate cache when user changes password.

EreMaijala committed:
Fix code style.

demiankatz reviewed (approved):
Thanks, @EreMaijala!
